### DISCUSSION TITLE
Question on IDM usage as ego policy

### DISCUSSION THREAD

**Discussion by AkshayNemade** [eleurent/highway-env#166] [General]
Hi :) ,

Can the ego vehicle be IDMvehcile? This feature would be usefull if the neighbouring vehicles are provided with real tracks from a dataset and  IDM is used maneuver amongst them.

Currently simply changing the self.vehicle = IDMVehicle() in abstract env doesn't work as it gives the error
AttributeError: 'IDMVehicle' object has no attribute 'SPEED_COUNT'

Any work around this?

**Comment by eleurent** [comment]
Hi @AkshayNemade
IDM is a longitudinal model, responsible for adjusting the vehicle velocity.
You could use it for the ego-vehicle, but then what would be the action space? Only lateral actions (lane change decisions or steering control), maybe?

For now, only two action spaces have been considered:
* a Discrete space, with 5 actions: IDLE, FASTER, SLOWER, LANE-RIGHT and LANE-LEFT.
* a Continuous space with 2 actions: throttle and steering

Thus, you get this error because the current Discrete action space requires being able to handle the FASTER/SLOWER actions, which involves a finite set of target speeds. These are not defined for IDM (which adjusts its acceleration continuously), so a work around would be to define a new action space with only lateral controls.

**Comment by AkshayNemade** [comment]
I am sorry for being not clear. I meant both IDM and MOBIL.. which IDM vehicle object use.
I see that the continuous actions are also passed in the env. So cant IDM+MOBIL, accelerations + steering controls passed internally?

**Comment by eleurent** [comment]
So you want the ego-vehicle to behave according to IDM+MOBIL, as for any other vehicle.
It is of course possible, but that would make highway-env a simulator, rather than a reinforcement learning environment: you cannot provide actions to control the simulation, which runs on its own.

You can achieve this by simply commenting the `vehicle.act(action)` line in `AbstractEnv`. This means that the action will be ignored when calling `env.step(action)`

**Reply by YuqiaoBai** [reply]
Hi, I need to set a IDMVehicle as ego vehicle. But I cant find the vehicle.act(action) line in AbstractEnv class.

**Reply by eleurent** [reply]
Hi @YuqiaoBai, sorry the code has changed a bit since then.
The action is now executed here:
https://github.com/eleurent/highway-env/blob/9d63973da854584fe51b00ccee7b24b1bf031418/highway_env/envs/common/abstract.py#L232

There are several implementations depending on the selected action type, see https://github.com/eleurent/highway-env/blob/master/highway_env/envs/common/action.py

**Reply by tahsinmostafa** [reply]
Hi, I was wondering if it is still possible to make the "ego-vehicle to behave according to IDM+MOBIL, as for any other vehicle" by just commenting out `self.action_type.act(action)` in `abstract.py`.

**Reply by eleurent** [reply]
Hi, yes I think it should still work.
